lib/Data/Time/Clock/POSIX.hs

```diff
@@ -1,84 +1,127 @@
---
-type POSIXTime = NominalDiffTime
+data POSIXTime = POSIXTime
+    { ptSeconds ::     {-# UNPACK #-} !Int64
+    , ptNanoSeconds :: {-# UNPACK #-} !Int64
+    }
+
+normalizePosix :: POSIXTime -> POSIXTime
+normalizePosix raw@(POSIXTime xs xn)
+    | xn < 0 || xn >= 1000000000 = POSIXTime (xs + q)  r
+    | otherwise                  = raw
+  where (q, r) = xn `divMod` 1000000000
+
+instance Eq POSIXTime where
+    rawx == rawy =
+        let POSIXTime xs xn = normalizePosix rawx
+            POSIXTime ys yn = normalizePosix rawy
+        in xs == ys && xn == yn
+
+instance Ord POSIXTime where
+    rawx `compare` rawy =
+        let POSIXTime xs xn = normalizePosix rawx
+            POSIXTime ys yn = normalizePosix rawy
+            os = compare xs ys
+        in if os == EQ then xn `compare` yn else os
+
+instance NFData POSIXTime where
+    rnf a = a `seq` ()
+
+posixToUTCTime :: POSIXTime -> UTCTime
+posixToUTCTime raw =
+    let POSIXTime s ns = normalizePosix raw
+        (d, s') = s `divMod` posixDayLength
+        ps = s' * 1000000000000 + ns * 1000 -- 'Int64' can hold ps in one day
+    in UTCTime (addDays (fromIntegral d) unixEpochDay)
+               (picosecondsToDiffTime (fromIntegral ps))
+
+posixDayLength :: Int64
+posixDayLength = 86400
 
 unixEpochDay :: Day
 unixEpochDay = ModifiedJulianDay 40587
 
-posixSecondsToUTCTime :: POSIXTime -> UTCTime
-posixSecondsToUTCTime i = let
-    (d,t) = divMod' i posixDayLength
- in UTCTime (addDays d unixEpochDay) (realToFrac t)
-
-utcTimeToPOSIXSeconds :: UTCTime -> POSIXTime
-utcTimeToPOSIXSeconds (UTCTime d t) =
- (fromInteger (diffDays d unixEpochDay) * posixDayLength) + min posixDayLength (realToFrac t)
-
--- | Get the current POSIX time from the system clock.
 getPOSIXTime :: IO POSIXTime
-
 #ifdef mingw32_HOST_OS
 -- On Windows, the equlvalent of POSIX time is "file time", defined as
 -- the number of 100-nanosecond intervals that have elapsed since
 -- 12:00 A.M. January 1, 1601 (UTC).  We can convert this into a POSIX
 -- time by adjusting the offset to be relative to the POSIX epoch.
 
 getPOSIXTime = do
-  FILETIME ft <- System.Win32.Time.getSystemTimeAsFileTime
-  return (fromIntegral (ft - win32_epoch_adjust) / 10000000)
-
-win32_epoch_adjust :: Word64
-win32_epoch_adjust = 116444736000000000
+    FILETIME ft <- System.Win32.Time.getSystemTimeAsFileTime
+    let (s, us) = (ft - win32_epoch_adjust) `divMod` 10000000
+    return (POSIXTime (fromIntegral s) (fromIntegral us * 1000))
+  where
+    win32_epoch_adjust :: Word64
+    win32_epoch_adjust = 116444736000000000
 
 #elif HAVE_CLOCK_GETTIME
+-- Use hi-res clock_gettime
 
--- Use hi-res POSIX time
-ctimespecToPosixSeconds :: CTimespec -> POSIXTime
-ctimespecToPosixSeconds (MkCTimespec (CTime s) ns) =
-    (fromIntegral s) + (fromIntegral ns) / 1000000000
-
-getPOSIXTime = liftM ctimespecToPosixSeconds getCTimespec
+getPOSIXTime = do
+    MkCTimespec (CTime s) (CLong ns) <- getCTimespec
+    return (POSIXTime s ns)
```

Comment:
Could you verify this? From http://hackage.haskell.org/package/base-4.9.0.0/docs/src/Foreign.C.Types.html#CTime :

    -- | Haskell type representing the C @time_t@ type.
    ARITHMETIC_TYPE(CTime,HTYPE_TIME_T)

As far as I can tell, `HTYPE_TIME_T` is equivalent to `time_t`.